File: football-gui-game/src/main/java/com/github/lipinskipawel/controller/HellController.java
package com.github.lipinskipawel.controller;

import com.github.lipinskipawel.board.engine.Board;
import com.github.lipinskipawel.board.engine.Boards;
import com.github.lipinskipawel.board.engine.Player;
import com.github.lipinskipawel.gui.DrawableFootballPitch;
import com.github.lipinskipawel.gui.RenderablePoint;
import kotlin.Unit;

import javax.swing.*;
import java.util.HashMap;
import java.util.Map;

import static com.github.lipinskipawel.board.engine.Player.FIRST;
import static com.github.lipinskipawel.board.engine.Player.SECOND;

final class HellController implements PitchController {

    private final DrawableFootballPitch drawableFootballPitch;
    private final Map<Player, Integer> tokenForPlayer;
    private GameFlowController gameFlowController;


    public HellController(final DrawableFootballPitch drawableFootballPitch) {
        this.drawableFootballPitch = drawableFootballPitch;
        this.gameFlowController = new GameFlowController(Boards.immutableBoard(), false);
        this.tokenForPlayer = new HashMap<>();
        this.tokenForPlayer.put(FIRST, 2);
        this.tokenForPlayer.put(SECOND, 2);
    }

    @Override
    public void leftClick(final RenderablePoint renderablePoint) {
        if (this.gameFlowController.isGameOver()) {
            return;
        }
        this.gameFlowController = this.gameFlowController.makeAMove(renderablePoint.getPosition());
        this.drawableFootballPitch.drawPitch(this.gameFlowController.board(), FIRST);
        this.gameFlowController.onWinner(this::winningMessage);
    }

    @Override
    public void rightClick(final RenderablePoint renderablePoint) {
        if (this.gameFlowController.isGameOver()) {
            return;
        }
        if (playerAllowedToUndo(this.gameFlowController.board())) {
            this.gameFlowController = this.gameFlowController.undoPlayerMove(
                    () -> {
                        final var dataObject = new QuestionService(new InMemoryQuestions())
                                .displayYesNoCancel("hell-mode");
                        new HerokuService().send(dataObject);
                        return null;
                    }
            );
            this.tokenForPlayer.compute(this.gameFlowController.player(), (key, val) -> val - 1);
            this.tokenForPlayer.compute(this.gameFlowController.player().opposite(), (key, val) -> val + 1);
            final var message =
                    "Player " + this.gameFlowController.player() +
                            " tokens : " + this.tokenForPlayer.get(this.gameFlowController.player()) + "\n" +
                            "Player " + this.gameFlowController.player().opposite() +
                            " tokens : " + this.tokenForPlayer.get(this.gameFlowController.player().opposite());
            JOptionPane.showMessageDialog(null, message);
        }
        this.drawableFootballPitch.drawPitch(this.gameFlowController.board(), FIRST);
    }

    private boolean playerAllowedToUndo(final Board board) {
        return this.tokenForPlayer.get(board.getPlayer()) > 0;
    }

    private Unit winningMessage(final Player winner) {
        JOptionPane.showMessageDialog(null, "Player " + winner + " won he game.");
        return null;
    }

    @Override
    public void tearDown() {
        this.gameFlowController = new GameFlowController(Boards.immutableBoard(), false);
        this.drawableFootballPitch.drawPitch(this.gameFlowController.board(), this.gameFlowController.player());
        this.tokenForPlayer.compute(FIRST, (currentValue, oldValue) -> 2);
        this.tokenForPlayer.compute(SECOND, (currentValue, oldValue) -> 2);
    }
}

File: src/main/kotlin/com/github/lipinskipawel/controller/HellController.kt
package com.github.lipinskipawel.controller

import com.github.lipinskipawel.board.engine.Board
import com.github.lipinskipawel.board.engine.Boards
import com.github.lipinskipawel.board.engine.Player
import com.github.lipinskipawel.game.GameFlowController
import com.github.lipinskipawel.gui.DrawableFootballPitch
import com.github.lipinskipawel.gui.RenderablePoint
import javax.swing.JOptionPane

internal class HellController(private val drawableFootballPitch: DrawableFootballPitch) : PitchController {
    private val tokenForPlayer: MutableMap<Player, Int>
    private var gameFlowController: GameFlowController

    init {
        gameFlowController = GameFlowController(Boards.immutableBoard(), false)
        tokenForPlayer = HashMap()
        tokenForPlayer[Player.FIRST] = 2
        tokenForPlayer[Player.SECOND] = 2
    }

    override fun leftClick(renderablePoint: RenderablePoint) {
        if (gameFlowController.isGameOver()) {
            return
        }
        gameFlowController = gameFlowController.makeAMove(renderablePoint.position)
        drawableFootballPitch.drawPitch(gameFlowController.board(), Player.FIRST)
        gameFlowController.onWinner { winner: Player -> winningMessage(winner) }
    }

    override fun rightClick(renderablePoint: RenderablePoint) {
        if (gameFlowController.isGameOver()) {
            return
        }
        if (playerAllowedToUndo(gameFlowController.board())) {
            gameFlowController = gameFlowController.undoPlayerMove { null }
            tokenForPlayer.compute(gameFlowController.player()) { key: Player?, `val`: Int? -> `val`!! - 1 }
            tokenForPlayer.compute(gameFlowController.player().opposite()) { key: Player?, `val`: Int? -> `val`!! + 1 }
            val message = """Player ${gameFlowController.player()} tokens : ${tokenForPlayer[gameFlowController.player()]}
Player ${gameFlowController.player().opposite()} tokens : ${tokenForPlayer[gameFlowController.player().opposite()]}"""
            JOptionPane.showMessageDialog(null, message)
        }
        drawableFootballPitch.drawPitch(gameFlowController.board(), Player.FIRST)
    }

    private fun playerAllowedToUndo(board: Board<Player>): Boolean {
        return tokenForPlayer[board.player]!! > 0
    }

    private fun winningMessage(winner: Player) {
        JOptionPane.showMessageDialog(null, "Player $winner won he game.")
    }

    override fun tearDown() {
        gameFlowController = GameFlowController(Boards.immutableBoard(), false)
        drawableFootballPitch.drawPitch(gameFlowController.board(), gameFlowController.player())
        tokenForPlayer.compute(Player.FIRST) { currentValue: Player?, oldValue: Int? -> 2 }
        tokenForPlayer.compute(Player.SECOND) { currentValue: Player?, oldValue: Int? -> 2 }
    }
}
